### ISSUE TITLE
Train (new) improved leg detector

### ISSUE THREAD

**Issue by zephirefaith** [CogRob/cogrob_ros#10]
Approach 1: Train a completely new leg detector of your own with dataset that we can control
1. Collect SICK 2D dataset
2. Find a way to annotate: Commercial or Crowdsourced? Planning on getting a large dataset for training deep NN
3. Do some sort of clustering on 2D points for hypothesis, or train that too?
4. Train a deep NN for recognizing legs of people

Approach 2: Current leg detectors give a lot of false positives, just add a second stage classifier which specializes in discrimination
1. Collect more SICK 2D data
2. Record *ALL* hypothesis from current leg-detector
3. Make a cascaded AdaBoost or SVM 2nd stage classifier which discriminates between true detections and false positives

**Comment by hicgt** [comment]
Lets talk about this. There are some easy ways to do this. The proposed approach is great but much of an overkill.

1. It is a Hokuyo sensor (not SICK)
2. If you use it in an open environment as the area with the windows the only objects in view will be legs so annotation / segmentation is easy.
3. Connected points (~1 in or less distance between neighboring points are clustered
4. Fit ellips to connected points for parameter estimation
5. Run ML on connected points and ellipse data to see which is the best method
6. Run leg doctor as robot runs through environment to get a sense for the number of distractors

I prefer to do this without deep NN as I would like to understand what we are learning. I think a CNN would be an overkill, but we can try both

Henrik


--
Henrik I. Christensen, Ph.D. Dr. Techn. h.c., IEEE & AAAS Fellow
Phone: +1 404 889 2500, LinkedIn<https://www.linkedin.com/in/henrikichristensen>, Web<http://www.hichristensen.net>, About.Me<https://about.me/hchristensen>, Twitter<http://twitter.com/hiskov>


On November 3, 2016 at 4:28:10 PM, Priyam Parashar (notifications@github.com<mailto:notifications@github.com>) wrote:

Approach 1: Train a completely new leg detector of your own with dataset that we can control

  1.  Collect SICK 2D dataset
  2.  Find a way to annotate: Commercial or Crowdsourced? Planning on getting a large dataset for training deep NN
  3.  Do some sort of clustering on 2D points for hypothesis, or train that too?
  4.  Train a deep NN for recognizing legs of people

Approach 2: Current leg detectors give a lot of false positives, just add a second stage classifier which specializes in discrimination

  1.  Collect more SICK 2D data
  2.  Record ALL hypothesis from current leg-detector
  3.  Make a cascaded AdaBoost or SVM 2nd stage classifier which discriminates between true detections and false positives

**Comment by BillWSY** [comment]
I think one of the factors we need to consider is whether the sensor gives us enough information to detect the legs. If we (human) were given the data from the sensors, what level of accuracy can we achieve?

If the algorithm can give us an even better result, we might need to find an alternative method to improve the accuracy (probably use information from other sensors).

**Comment by zephirefaith** [comment]
@hicgt  Firstly, I like the idea of getting our own dataset regardless of whatever technique we use. Since we are aiming for a complete, robust platform, it'll be a good idea to make it robust from bottom-up (specially since we have seen off-the-shelf stuff is lacking).
And if I am not wrong, what you're suggesting is a two-stage approach, right? First make the basic detector from our own dataset, then find the distractors and either make another classifier or some other solution depending upon the results.
I wanted to try out neural nets mostly because I wanted to get some hands on experience with them + looks good on resume for internships and such 😁 

@BillWSY You're right about that. We do plan on fusing these measurements with more sensors, for better tracking, for "seeding" legs with faces or vice versa, etc. But I want to do this first-hand at least once, just to see how good a leg detector can we make? My experience with off-the-shelf codes hasn't been amazing. If we can push the envelope with this component...why not?

**Comment by BillWSY** [comment]
I agree. Our brain is brilliant at coming up with new ideas when we do work -- even if we were replicating other people's work. With the infrastructure we have, I think we can easily many ideas -- with or without NN. Let me know if there is anything that I can help with.
